### PR TITLE
Basic synchronization logic of sync manager

### PR DESCRIPTION
**Pull request by tomasMizera**
PR adds logic to `SynchronizationManager` class which is now usable for syncing.

It also adds another class called `AutosyncController` which will be responsible for autosync logic (listen on data change in layers, error handling and incremental sync).

If `autosync` feature will be _enabled_, `SynchronizationManager` will instantiate `AutosyncController` and assign it currently opened project. When project will change / autosync will be _disabled_, `SynchronizationManager` will reset its `AutosyncController` / remove it altogether.

There were also few renaming:
 - `project upload` renamed to -> `project push`
 -  `project update` renamed to -> `project pull`
 -  `synchronization controller` renamed to -> `synchronization manager`

PR also adds autosync entry to persistent storage (AppSettings).

### CODE REVIEW

File: app/synchronizationmanager.cpp [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+  QObject::connect( mBackend, &MerginApi::syncProjectFinished, this, &SynchronizationManager::syncProjectFinished );
+  QObject::connect( mBackend, &MerginApi::syncProjectStatusChanged, this, &SynchronizationManager::syncProjectStatusChanged );
+}
+
+SynchronizationManager::~SynchronizationManager()
+{
+
+}
+
+void SynchronizationManager::syncProject( const Project &project, bool withAuth )
```
Comment (varmar05) [inline app/synchronizationmanager.cpp:26]: it would be nice if project 'knew' whether auth is needed or not

File: app/appsettings.cpp [hunk 2]
```diff
@@ -30,6 +30,7 @@ AppSettings::AppSettings( QObject *parent ): QObject( parent )
   QString savedAppVersion = settings.value( QStringLiteral( "appVersion" ), QStringLiteral() ).toString();
   bool legacyFolderMigrated = settings.value( QStringLiteral( "legacyFolderMigrated" ), false ).toBool();
   QString activeProviderId = settings.value( QStringLiteral( "activePositionProviderId" ) ).toString();
+  bool autosync = settings.value( QStringLiteral( "autosyncAllowed" ) ).toBool();
```
Comment (PeterPetrik) [inline app/appsettings.cpp:33]: default should be false?

File: app/autosynccontroller.h [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+{
+    Q_OBJECT
+
+    Q_PROPERTY( SyncStatus syncStatus READ syncStatus NOTIFY syncStatusChanged )
+
+  public:
+
+    enum SyncStatus
+    {
+      Synced = 0,
```
Comment (PeterPetrik) [inline app/autosynccontroller.h:30]: Synced == WaitingForNextSync? or means it will not try to sync anymore?

**Review by varmar05 — APPROVED** [review]

File: app/autosynccontroller.h [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+
+  public slots:
+
+    void handleSyncFinished();
+
+  private:
+
+    SyncStatus mSyncStatus = Synced;
+
+    Project *mActiveProject = nullptr;
```
Comment (PeterPetrik) [inline app/autosynccontroller.h:61]: I assume it is not owned by this class? Considering switching to some smart pointers? detto for mActiveQgsProject (not sure how it is else where in the code)

File: app/projectsmodel.cpp [hunk 5]
```diff
@@ -635,3 +621,18 @@ ProjectsModel::ProjectModelTypes ProjectsModel::modelType() const
...
+}
+
+void ProjectsModel::setSyncManager( SynchronizationManager *newSyncManager )
+{
+  if ( mSyncManager == newSyncManager )
+    return;
+  mSyncManager = newSyncManager;
+  emit syncManagerChanged( mSyncManager );
+
+  initializeProjectsModel();
```
Comment (PeterPetrik) [inline app/projectsmodel.cpp:637]: not sure if this call should be in setter.... Create internal connection for syncManagerChanged signal?

File: app/projectsmodel.h [hunk 6]
```diff
@@ -15,6 +15,7 @@
 
 #include "project.h"
 #include "merginapi.h"
+#include "synchronizationmanager.h"
```
Comment (PeterPetrik) [inline app/projectsmodel.h:18]: forward declaration?

File: app/synchronizationmanager.cpp [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+SynchronizationManager::SynchronizationManager( MerginApi *backend, QObject *parent )
+  : QObject( parent )
+  , mAutosyncController( nullptr )
+  , mBackend( backend )
+{
+  QObject::connect( mBackend, &MerginApi::syncProjectFinished, this, &SynchronizationManager::syncProjectFinished );
+  QObject::connect( mBackend, &MerginApi::syncProjectStatusChanged, this, &SynchronizationManager::syncProjectStatusChanged );
+}
+
+SynchronizationManager::~SynchronizationManager()
```
Comment (PeterPetrik) [inline app/synchronizationmanager.cpp:21]: = default

File: app/synchronizationmanager.cpp [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "synchronizationmanager.h"
+
+SynchronizationManager::SynchronizationManager( MerginApi *backend, QObject *parent )
+  : QObject( parent )
+  , mAutosyncController( nullptr )
```
Comment (PeterPetrik) [inline app/synchronizationmanager.cpp:14]: initialize in header to nullptr

File: app/synchronizationmanager.cpp [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+  }
+  else if ( project.mergin->status == ProjectStatus::Modified )
+  {
+    mBackend->pushProject( project.mergin->projectNamespace, project.mergin->projectName );
+  }
+}
+
+void SynchronizationManager::stopProjectSync( const QString &projectFullname )
+{
+  Transactions t = mBackend->transactions();
```
Comment (PeterPetrik) [inline app/synchronizationmanager.cpp:45]: is this deep copy?

File: core/project.cpp [hunk 10]
```diff
@@ -51,7 +51,7 @@ MerginProject *MerginProject::clone() const
   return me;
 }
 
-ProjectStatus::Status ProjectStatus::projectStatus( const std::shared_ptr<Project> project )
```
Comment (PeterPetrik) [inline core/project.cpp:54]: why we do not use smart pointers anymore?

**Comment by PeterPetrik** [comment]
looks like I reviewed a minute late :(

File: app/autosynccontroller.h (hunk 3 above)
Comment (tomasMizera) [reply app/autosynccontroller.h:30]: I'd say it is a state when sync succeeded and there are no pending changes. These states, though, might change with implementation of the controller. It was more like a first try

**Pull request merged**